### PR TITLE
Refresh Python Engine when IronPython loaded

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

Per @aparajit-pratap 's suggestion

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@DynamoDS/dynamo 

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

**Comment by aparajit-pratap** [comment]
@QilongTang, I guess the best and easiest way to test this is after publishing the latest package. I wasn't able to test it very well without a package install workflow. Thanks!

File: src/Libraries/PythonNodeModels/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -12,4 +12,5 @@
 [assembly: Guid("a0debf47-0cf2-46e8-90b9-0c3d9b808fd7")]
 // Remove this when PythonEngineSelector become public
 [assembly: InternalsVisibleTo("DynamoPythonTests")]
-[assembly: InternalsVisibleTo("DynamoCoreWpfTests")]
\ No newline at end of file
+[assembly: InternalsVisibleTo("DynamoCoreWpfTests")]
+[assembly: InternalsVisibleTo("IronPythonExtension")]
```
Comment (mmisol) [inline src/Libraries/PythonNodeModels/Properties/AssemblyInfo.cs:16]: I don't think this helps in the cause of externalizing the extension. We should use an event and a proper API.

Comment (mjkkirschner) [reply src/Libraries/PythonNodeModels/Properties/AssemblyInfo.cs:16]: `RequestUpdateLoadedPythonEngines`?

Comment (aparajit-pratap) [reply src/Libraries/PythonNodeModels/Properties/AssemblyInfo.cs:16]: For this Core would need to be aware of the event on the extension, in order to register its scan method with it, which does not seem right. An extension is and should be aware of Core, not the other way round.

**Comment by aparajit-pratap** [comment]
> An extension that is going to become external should not have access to internals of a core library.

True, but in this case, we need to make an exception as this extension is maintained by us. I think we have made similar exceptions for other extensions that are written by the Dynamo team.

**Comment by QilongTang** [comment]
> > An extension that is going to become external should not have access to internals of a core library.
> 
> True, but in this case, we need to make an exception as this extension is maintained by us. I think we have made similar exceptions for other extensions that are written by the Dynamo team.

@mmisol @mjkkirschner I would like to emphasize, the Engine Selector APIs were not finalized, that is the reason why we make them internal visible to other binaries for now so we do not have to obsolete API soon. But since you insist, I've made them public with the new commit, PTAL

**Comment by QilongTang** [comment]
@mmisol @mjkkirschner I think since our API scope is not clear yet, it is still part of our API now. However it is ready in its current state. We would just need to make proper changes when we need to implement dynamic loading, but if we do not get to the dynamic loading task, the current state is still good to use for extension or 3rd party devs.

**Comment by mjkkirschner** [comment]
Agree with @mmisol's final statement here:

> I guess I feel the last change rushes an API that may not be what we want. If that's the case, the outcome would be worse than without these changes, because at least we would not have made anything public.
Finalizing the API was the idea I had in my mind, but if this won't be done before this PR gets merged, I'd rather take the previous change than this.

It's not clear we can finalize the API until we finalize the dynamic loading - so my vote would be to merge this in a temporary internal state and move on to that to reasses.

**Comment by QilongTang** [comment]
@aparajit-pratap Are you fine with me reverting the last commit then?

**Comment by aparajit-pratap** [comment]
I'd like to point out that the `PythonEngineSelector` is also being used by the PythonMigration view extension now in https://github.com/DynamoDS/Dynamo/pull/10869 and since it was initially made internal, I had to make it visible to the python migration view extension as well. I think the crux of the matter is that these extensions that we have created recently (like the IronPython and Python migration extensions) are more internal in their core functionality and that it is unlikely that third party devs would be creating extensions like these in the future and so in that sense, these are in a class apart from general-purpose extensions and view extensions.

**Comment by QilongTang** [comment]
> I'd like to point out that the `PythonEngineSelector` is also being used by the PythonMigration view extension now in #10869 and since it was initially made internal, I had to make it visible to the python migration view extension as well. I think the crux of the matter is that these extensions that we have created recently (like the IronPython and Python migration extensions) are more internal in their core functionality and that it is unlikely that third party devs would be creating extensions like these in the future and so in that sense, these are in a class apart from general-purpose extensions and view extensions.

I agree

**Comment by QilongTang** [comment]
@aparajit-pratap @mmisol @mjkkirschner Reverted. Let me know

**Pull request merged**